std/compress/lzss/snark.go

```diff
@@ -10,43 +10,44 @@ import (
 	"github.com/consensys/gnark/std/lookup/logderivlookup"
 )
 
+// TODO Provide option for c to be in sizes other than bytes
+
 // Decompress decompresses c into d using dict as the dictionary
 // which must come pre "augmented"
 // it is on the caller to ensure that the dictionary is correct; in particular it must consist of bytes. Decompress does not check this.
 // it is recommended to pack the dictionary using compress.Pack and take a MiMC checksum of it.
 // d will consist of bytes
 // It returns the length of d as a frontend.Variable; if the decompressed stream doesn't fit in d, dLength will be "-1"
-func Decompress(api frontend.API, c []frontend.Variable, cLength frontend.Variable, d, dict []frontend.Variable, level lzss.Level) (dLength frontend.Variable, err error) {
+func Decompress(api frontend.API, c []frontend.Variable, cLength frontend.Variable, d, dict []frontend.Variable) (dLength frontend.Variable, err error) {
 
 	api.AssertIsLessOrEqual(cLength, len(c)) // sanity check
 
 	// size-related "constants"
-	wordNbBits := int(level)
-	shortBackRefType, longBackRefType, dictBackRefType := lzss.InitBackRefTypes(len(dict), level) // init the dictionary and backref types; only needed for the constants below
-	shortBrNbWords := int(shortBackRefType.NbBitsBackRef) / wordNbBits
-	longBrNbWords := int(longBackRefType.NbBitsBackRef) / wordNbBits
-	dictBrNbWords := int(dictBackRefType.NbBitsBackRef) / wordNbBits
-	byteNbWords := uint(8 / wordNbBits)
+	shortBackRefType := lzss.NewShortBackrefType()
+	dynamicBackRefType := lzss.NewDynamicBackrefType(len(dict), 0)
 
 	// check header: version and compression level
 	const (
 		sizeHeader = 3
-		version    = 0
+		version    = 1
 	)
+	api.AssertIsLessOrEqual(sizeHeader, len(c))
 	api.AssertIsEqual(c[0], version/256)
 	api.AssertIsEqual(c[1], version%256)
-	fileCompressionMode := c[2]
-	api.AssertIsEqual(api.Mul(fileCompressionMode, fileCompressionMode), api.Mul(fileCompressionMode, wordNbBits)) // if fcm!=0, then fcm=wordNbBits
-	decompressionNotBypassed := api.Sub(1, api.IsZero(fileCompressionMode))
+	decompressionBypassed := c[2]
+	api.AssertIsBoolean(decompressionBypassed)
+	if len(c) == 3 {
+		return 0, nil
+	}
 
 	// check that the input is in range and convert into small words
 	rangeChecker := internal.NewRangeChecker(api)
 
 	bytes := make([]frontend.Variable, len(c)-sizeHeader+1)
 	copy(bytes, c[sizeHeader:])
-	bytes[len(bytes)-1] = 0                                             // pad with a zero to avoid out of range errors
-	c, bytes = rangeChecker.BreakUpBytesIntoWords(wordNbBits, bytes...) // from this point on c is in words
-	cLength = api.Mul(api.Sub(cLength, sizeHeader), 8/wordNbBits)       // one constraint; insignificant impact anyway
+	bytes[len(bytes)-1] = 0                                    // pad with a zero to avoid out of range errors
+	c, bytes = rangeChecker.BreakUpBytesIntoWords(1, bytes...) // from this point on c is in bits
```

Comment:
can the code linked to this be simplified ? (i.e now that we assume 8 bits per word all the time; )

Reply:
I think so. We can just use `api.ToBinary` but I'd like to wait for the next version because I'm planning some input granularity shenanigans where this may come in handy.